pkg/ruler/remotequerier.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	req := httpgrpc.HTTPRequest{
+		Method: http.MethodPost,
+		Url:    q.promHTTPPrefix + readEndpointPath,
+		Body:   snappy.Encode(nil, data),
+		Headers: []*httpgrpc.Header{
+			{Key: textproto.CanonicalMIMEHeaderKey("Content-Encoding"), Values: []string{"snappy"}},
+			{Key: textproto.CanonicalMIMEHeaderKey("Accept-Encoding"), Values: []string{"snappy"}},
+			{Key: textproto.CanonicalMIMEHeaderKey("Content-Type"), Values: []string{"application/x-protobuf"}},
+			{Key: textproto.CanonicalMIMEHeaderKey("User-Agent"), Values: []string{userAgent}},
+			{Key: textproto.CanonicalMIMEHeaderKey("X-Prometheus-Remote-Read-Version"), Values: []string{"0.1.0"}},
+		},
+	}
+
+	for _, mdw := range q.middlewares {
+		if err := mdw(ctx, &req); err != nil {
+			return nil, err
+		}
+	}
+
+	resp, err := q.client.Handle(ctx, &req)
+	if err != nil {
+		level.Warn(log).Log("msg", "failed to perform remote read", "err", err, "qs", query)
+		return nil, err
+	}
+	level.Debug(log).Log("msg", "remote read successfully performed", "qs", query)
+
+	if resp != nil && resp.Code/100 != 2 {
+		return nil, fmt.Errorf("unexpected response status code %d: %s", resp.Code, string(resp.Body))
+	}
+
+	uncompressed, err := snappy.Decode(nil, resp.Body)
+	if err != nil {
+		return nil, errors.Wrap(err, "error reading response")
+	}
+	var rdResp prompb.ReadResponse
+
+	err = proto.Unmarshal(uncompressed, &rdResp)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to unmarshal response body")
+	}
+
+	if len(rdResp.Results) != 1 {
+		return nil, errors.Errorf("responses: want %d, got %d", 1, len(rdResp.Results))
+	}
+	return rdResp.Results[0], nil
+}
+
+// Query performs a query for the given time.
+func (q *RemoteQuerier) Query(ctx context.Context, query string, ts time.Time) (model.ValueType, json.RawMessage, error) {
+	log, ctx := spanlogger.NewWithLogger(ctx, q.logger, "remotequerier.Querier.Query")
+	defer log.Span.Finish()
+
+	args := make(url.Values)
+	args.Set("query", query)
+	if !ts.IsZero() {
+		args.Set("time", formatQueryTime(ts))
+	}
+	body := []byte(args.Encode())
+
+	req := httpgrpc.HTTPRequest{
+		Method: http.MethodPost,
+		Url:    q.promHTTPPrefix + queryEndpointPath,
+		Body:   body,
+		Headers: []*httpgrpc.Header{
+			{Key: textproto.CanonicalMIMEHeaderKey("User-Agent"), Values: []string{userAgent}},
+			{Key: textproto.CanonicalMIMEHeaderKey("Content-Type"), Values: []string{mimeTypeFormPost}},
+			{Key: textproto.CanonicalMIMEHeaderKey("Content-Length"), Values: []string{strconv.Itoa(len(body))}},
+		},
+	}
+
+	for _, mdw := range q.middlewares {
+		if err := mdw(ctx, &req); err != nil {
+			return model.ValNone, nil, err
+		}
+	}
+
+	resp, err := q.client.Handle(ctx, &req)
+	if err != nil {
+		level.Warn(log).Log("msg", "failed to remotely evaluate query expression", "err", err, "qs", query, "tm", ts)
+		return model.ValNone, nil, err
+	}
+	level.Debug(log).Log("msg", "query expression successfully evaluated", "qs", query, "tm", ts)
```

Comment:
Should this log be at least after we've checked the response code?

Reply:
totally agree! nice catch.